User provided Rest client @RegisterProvider does not work

I am running into a problem for the following in native
```
@Path("/")
@RegisterRestClient
@RegisterProvider(GitHubApiExceptionMapper.class)
public interface GitHubApiService {

```
which gives
```
Caused by: java.lang.InstantiationException: Type `org.overbaard.review.tool.rest.client.github.GitHubApiExceptionMapper` can not be instantiated reflectively as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.
	at java.lang.Class.newInstance(DynamicHub.java:678)
	at io.quarkus.smallrye.restclient.runtime.RestClientBuilderImpl.newInstanceOf(RestClientBuilderImpl.java:342)
	... 69 more
```
Kabir Khan: I tried adding a default constructor but get the same error. How do I add the constructor to the native image? I think I know how to do it in an extension (https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/UUIDGenerator.20(native)/near/161981147) but don't see how to do it for user-provided code

**Ken Finnigan: please raise an issue, it likely means we need to register constructors of classes in that annotation for native image**

--- a/extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java
+++ b/extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java
@@ -17,8 +17,11 @@
 package io.quarkus.smallrye.restclient.deployment;
 
 import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
@@ -29,6 +32,8 @@ import javax.ws.rs.ext.Providers;
 
 import org.apache.commons.logging.impl.Jdk14Logger;
 import org.apache.commons.logging.impl.LogFactoryImpl;
+import org.eclipse.microprofile.rest.client.annotation.RegisterProvider;
+import org.eclipse.microprofile.rest.client.annotation.RegisterProviders;
 import org.eclipse.microprofile.rest.client.ext.DefaultClientHeadersFactoryImpl;
 import org.eclipse.microprofile.rest.client.inject.RestClient;
 import org.jboss.jandex.AnnotationInstance;
@@ -73,6 +78,9 @@ class SmallRyeRestClientProcessor {
     private static final DotName REST_CLIENT = DotName.createSimple(RestClient.class.getName());
 
     private static final DotName PATH = DotName.createSimple(Path.class.getName());
+
+    private static final DotName REGISTER_PROVIDER = DotName.createSimple(RegisterProvider.class.getName());
+    private static final DotName REGISTER_PROVIDERS = DotName.createSimple(RegisterProviders.class.getName());
 
     private static final String PROVIDERS_SERVICE_FILE = "META-INF/services/" + Providers.class.getName();
 
@@ -216,13 +224,26 @@ class SmallRyeRestClientProcessor {
 
     @BuildStep
     void registerProviders(BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
-            JaxrsProvidersToRegisterBuildItem jaxrsProvidersToRegisterBuildItem) {
+            JaxrsProvidersToRegisterBuildItem jaxrsProvidersToRegisterBuildItem,
+            CombinedIndexBuildItem combinedIndexBuildItem) {
         RestClientBuilderImpl.REGISTER_BUILTIN_PROVIDERS = jaxrsProvidersToRegisterBuildItem.useBuiltIn();
         RestClientBuilderImpl.PROVIDERS_TO_REGISTER = String.join(",", jaxrsProvidersToRegisterBuildItem.getProviders());
 
         // register the providers for reflection
         for (String providerToRegister : jaxrsProvidersToRegisterBuildItem.getProviders()) {
             reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, providerToRegister));
+        }
+
+        // now we register all values of @RegisterProvider for constructor reflection
+
+        IndexView index = combinedIndexBuildItem.getIndex();
+        List<AnnotationInstance> allInstances = new ArrayList<>(index.getAnnotations(REGISTER_PROVIDER));
+        for (AnnotationInstance annotation : index.getAnnotations(REGISTER_PROVIDERS)) {
+            allInstances.addAll(Arrays.asList(annotation.value().asNestedArray()));
+        }
+        for (AnnotationInstance annotationInstance : allInstances) {
+            reflectiveClass
+                    .produce(new ReflectiveClassBuildItem(false, false, annotationInstance.value().asClass().toString()));
         }
     }
 

